src/gl/index_buffer.js

```diff
@@ -1,35 +1,35 @@
 // @flow
 
-import type {SerializedStructArray} from '../util/struct_array';
+import type {TriangleIndexArray, LineIndexArray} from '../data/index_array_type';
 
 class IndexBuffer {
-    arrayBuffer: ?ArrayBuffer;
-    length: number;
     gl: WebGLRenderingContext;
-    buffer: ?WebGLBuffer;
+    buffer: WebGLBuffer;
 
-    constructor(array: SerializedStructArray) {
-        this.arrayBuffer = array.arrayBuffer;
-        this.length = array.length;
-    }
+    constructor(gl: WebGLRenderingContext, array: TriangleIndexArray | LineIndexArray) {
+        this.gl = gl;
+        this.buffer = gl.createBuffer();
 
-    bind(gl: WebGLRenderingContext) {
-        if (!this.buffer) {
-            this.gl = gl;
-            this.buffer = gl.createBuffer();
-            gl.bindBuffer(gl.ELEMENT_ARRAY_BUFFER, this.buffer);
-            gl.bufferData(gl.ELEMENT_ARRAY_BUFFER, this.arrayBuffer, gl.STATIC_DRAW);
-
-            // dump array buffer once it's bound to gl
-            this.arrayBuffer = null;
-        } else {
-            gl.bindBuffer(gl.ELEMENT_ARRAY_BUFFER, this.buffer);
+        if (gl.extVertexArrayObject === undefined) {
+            (gl: any).extVertexArrayObject = gl.getExtension("OES_vertex_array_object");
+        }
+        if (gl.extVertexArrayObject) {
+            (gl: any).extVertexArrayObject.bindVertexArrayOES(null);
```

Comment:
Why was this `extVertexArrayObject` stuff not necessary in this class before this change?  (I haven't had occasion to learn what "VAO"s are all about)

Reply:
Vertex array objects are essentially a mechanism to group together some pieces of OpenGL state such that that state can be switched more efficiently, all at once, by changing the bound VAO, rather than piece by piece. One of the pieces of VAO state is the bound element (index) buffer. That means that if you call `gl.bindBuffer(gl.ELEMENT_ARRAY_BUFFER, ...)`, you're _modifying the currently bound VAO_. There's no way to bind an element buffer without modifying some VAO.

One way to expose this reality would be to expose element buffer binding only through a wrapper method on our `VertexArrayObject` class, and never call `gl.bindBuffer(gl.ELEMENT_ARRAY_BUFFER, ...)` directly. But this is awkward because when we're creating the index buffer in the first place, there's not an obvious single `VertexArrayObject` instance available to use. (In fact, when the number of vertices being indexed exceeds `UINT16_MAX`, the created index buffer will eventually be bound in _multiple_ VAOs, one for each segment.) And we certainly don't want to mutate whatever VAO happens to currently be bound. So instead, we make sure we're modifying a special "global" VAO -- in OpenGL, this is done by binding VAO ID 0, in WebGL by binding `null`.

Where the VAO extension isn't available, you can think of it as an implementation where this "global" VAO is the only VAO that can exist -- and that's in fact [how we unify the API](https://github.com/mapbox/mapbox-gl-native/blob/6cbeeff6898211d3c3955e34503358e8b2ee127b/src/mbgl/gl/context.cpp#L304-L306) in native.

I'll add a comment here to explain.

Reply:
Oh, to answer why wasn't this necessary before: because we created index buffers on the fly, when (one of) the relevant VAOs was guaranteed to be bound.